File: src/servlets/AdminEditRoll.java
import java.io.*;
import java.util.*;

import javax.servlet.*;
import javax.servlet.http.*;

import freemarker.template.*;

import except.*;
import model.ctrl.*;
import model.entity.*;

public class AdminEditRoll extends HttpServlet {
  static  String templateDir = "WEB-INF/templates";
  private Configuration  cfg;

  public void init() {
    // Prepare the FreeMarker configuration;
    // - Load templates from the WEB-INF/templates directory of the Web app.
    //
    cfg = new Configuration();
    cfg.setServletContextForTemplateLoading( getServletContext(), "WEB-INF/templates");
  }
  
  public void doGet( HttpServletRequest  req, HttpServletResponse res ) throws ServletException, IOException {
 	Template resultTemplate = null;
    BufferedWriter out = null;
    User u = null;
    String resultTemplateName = "AdminEditRoll.ftl",
    		error = null;

    HttpSession session = req.getSession();
	u = (User) session.getAttribute( "user" );
    
    // Setup the data-model
    //
    Map vars = new HashMap();

    // Build the data-model
    //
    if(u == null) resultTemplateName = "Login.ftl";
    
    // Load templates from the WEB-INF/templates directory of the Web app.
    //
    try {
      resultTemplate = cfg.getTemplate( resultTemplateName );
    }catch (IOException e) {
      throw new ServletException("Can't load template in: " + templateDir + ": " + e.toString());
    }

    // Prepare the HTTP response:
    // - Use the charset of template for the output
    // - Use text/html MIME-type
    //
    out = new BufferedWriter(new OutputStreamWriter( res.getOutputStream(), resultTemplate.getEncoding() ));
    
    res.setContentType("text/html; charset=" + resultTemplate.getEncoding());

    // Merge the data-model and the template
    //
    try {
      resultTemplate.process( vars, out );
      out.flush();
    }catch (TemplateException e) {
      throw new ServletException( "Error while processing template", e);
    }
    out.close();
  }

  public void doPost( HttpServletRequest  req, HttpServletResponse res ) throws ServletException, IOException {
  	PrintWriter out = res.getWriter();
    res.setContentType("text/plain; charset=UTF-8");
    
    String dept = req.getParameter( "dept" );
    String course = req.getParameter( "course" );
    String section = req.getParameter( "section" );
    String selected = req.getParameter( "selected" );
    String enroll = req.getParameter( "enroll" );
    String student = req.getParameter( "student" );
    String unenroll = req.getParameter( "unenroll" );
    String search = req.getParameter( "search" );
    
    HttpSession session = req.getSession();
	User u = (User) session.getAttribute( "user" );
    
    StringBuffer content = new StringBuffer();
    
	content.append("{\n");	content.append("  \"results\": {\n");	content.append("    \"bindings\": [\n\n");
    
    if(dept != null){
    	try{
	    	for(Iterator iter = CtrlRegistration.getDepartments(); iter.hasNext();){
	    		Department d = (Department) iter.next();
	    		content.append("      {\n");				content.append("        \"department\": { \"id\": \""+d.getId()+"\", \"name\": \""+d.getName()+"\"  }\n");				content.append("      } ,\n");
	    	}
	    	content.setCharAt(content.length()-2, ' ');
    	}catch(PortsException e){}
    }else if(course != null){
    	try{
	    	for(Iterator iter = CtrlRegistration.getCourses(Integer.parseInt(course)); iter.hasNext();){
	    		Course c = (Course) iter.next();
	    		content.append("      {\n");				content.append("        \"course\": { \"id\": \""+c.getId()+"\", \"dept\": \""+
											c.getDeptId()+"\", \"name\": \""+c.getName()+"\", \"desc\": \""+c.getDesc()+"\" }\n");				content.append("      } ,\n");
	    	}
	    	content.setCharAt(content.length()-2, ' ');
	    }catch(PortsException e){}
    }else if(search != null && section != null){
    	try{
	    	for(Iterator iter = CtrlAdminEditRoll.getStudents(Integer.parseInt(section), search); iter.hasNext();){
	    		Student s = (Student) iter.next();
	    		content.append("      {\n");				content.append("        \"student\": { \"id\": \""+s.getId()+"\", \"name\":\""+
										s.getName()+"\", \"type\": \""+s.getType()+"\", \"address\": \""+
										s.getAddress()+"\"  }\n");				content.append("      } ,\n");
	    	}
	    	content.setCharAt(content.length()-2, ' ');
	    }catch(PortsException e){}
    }else if(section != null){
    	try{
	    	for(Iterator iter = CtrlRegistration.getSections(Integer.parseInt(section)); iter.hasNext();){
	    		Section s = (Section) iter.next();
	    		Faculty f = CtrlRegistration.getFaculty(s.getFacultyId());
	    		String facName = f.getName();
	    		content.append("      {\n");				content.append("        \"section\": { \"id\": \""+s.getId()+"\", \"course\":\""+
										s.getCourseId()+"\", \"faculty\": \""+facName+"\", \"days\": \""+
										s.getDays()+"\", \"period\": \""+s.getPeriod()+"\"  }\n");				content.append("      } ,\n");
	    	}
	    	content.setCharAt(content.length()-2, ' ');
	    }catch(PortsException e){}
    }else if(selected != null){
    	try{
	    	for(Iterator iter = CtrlAdminEditRoll.getStudentsOnRoll(Integer.parseInt(selected)); iter.hasNext();){
	    		Student s = (Student) iter.next();
	    		content.append("      {\n");				content.append("        \"student\": { \"id\": \""+s.getId()+"\", \"name\":\""+
										s.getName()+"\", \"type\": \""+s.getType()+"\", \"address\": \""+
										s.getAddress()+"\"  }\n");				content.append("      } ,\n");
	    	}
	    	content.setCharAt(content.length()-2, ' ');
    	}catch(PortsException e){}
    }else if(enroll != null && student != null){
    	content.append("      {\n");
    	try{
	    	int sectionId = Integer.parseInt(enroll);
	    	int studentId = Integer.parseInt(student);
	    	CtrlRegistration.enrollStudent(sectionId, studentId);
	    	content.append("        \"success\": \"true\"\n");
	    }catch(PortsException e){content.append("        \"success\": \"false\"\n");}
	    content.append("      } ,\n");
    }else if(unenroll != null && student != null){
    	content.append("      {\n");
    	try{
	    	int sectionId = Integer.parseInt(unenroll);
	    	int studentId = Integer.parseInt(student);
	    	CtrlRegistration.unEnrollStudent(sectionId, studentId);
	    	content.append("        \"success\": \"true\"\n");
	    }catch(PortsException e){content.append("        \"success\": \"false\"\n");}
	    content.append("      } ,\n");
    }
    content.append("    ]\n");	content.append("  }\n");	content.append("}\n");
    out.println(content.toString());
  }

}

File: web/js/adminEditRoll.js
var base = "AdminEditRoll?";

function getDepartments(){
	$("#status")[0].innerHTML = "";
	var url=base+"dept";
	
	var callback = function(data) {
		var json = parseJSON(data);
		var b = json["results"]["bindings"];
		var html = "<option value='-1'>--</option>";
		for(var i=0; i<b.length; i++){
			var dept = b[i]["department"];
			html += "<option value='"+dept.id+"'>"+dept.name+"</option>\n";
		}
		$("#dept")[0].innerHTML = html;
	};
	
	$.post(url, callback);
}

function getCourses(deptId){
	$("#status")[0].innerHTML = "";
	if(deptId != -1){
		var url=base+"course="+deptId;
		
		var callback = function(data) {	
			var json = parseJSON(data);
			var b = json["results"]["bindings"];
			var html = "<option value='-1'>--</option>";
			for(var i=0; i<b.length; i++){
				var course = b[i]["course"];
				html += "<option value='"+course.id+"'>"+course.name+"</option>\n";
			}
			$("#course")[0].innerHTML = html;
		};
		
		$.post(url, callback);
	}
}

function getSections(courseId){
	$("#status")[0].innerHTML = "";
	if(courseId != -1){
		var url=base+"section="+courseId;
		
		var callback = function(data) {
			var json = parseJSON(data);
			var b = json["results"]["bindings"];
			var html = "<option value='-1'>--</option>";
			for(var i=0; i<b.length; i++){
				var section = b[i]["section"];
				html += "<option value='"+section.id+"'>"+section.days+"-"+section.period+"</option>\n";
			}
			$("#section")[0].innerHTML = html;
		};
		
		$.post(url, callback);
	}
}

function getEnrolledStudents(sectionId){
	var url=base+"selected="+sectionId;
	
	var callback = function(data) {
		var json = parseJSON(data);
		var b = json["results"]["bindings"];
		
		var html = "<table><tbody><tr><th>Id</th><th>Name</th><th>Type</th><th>Address</th></tr>";
		for(var i=0; i<b.length; i++){
			var student = b[i]["student"];
			html += "<tr><td>"+student.id+"</td><td>"+student.name+"</td><td>"+
						student.type+"</td><td>"+student.address+"</td>"+
						"<td><input type='button' value='Remove' onclick='unEnroll("+student.id+")'/></td></tr>\n";
		}
		html +="</table>";
		$("#enrolled")[0].innerHTML = html;
	};
	
	$.post(url, callback);
}

function searchStudents(stuName){
	var section = $("#section")[0];
	if(section.selectedIndex != -1 && section.options[section.selectedIndex].value != -1){
		var url=base+"search="+stuName+"&section="+section.options[section.selectedIndex].value;
		
		var callback = function(data) {
			var json = parseJSON(data);
			var b = json["results"]["bindings"];
			
			var html = "<table><tbody><tr><th>Id</th><th>Name</th><th>Type</th><th>Address</th></tr>";
			for(var i=0; i<b.length; i++){
				var student = b[i]["student"];
				html += "<tr><td>"+student.id+"</td><td>"+student.name+"</td><td>"+
				student.type+"</td><td>"+student.address+"</td>"+
				"<td><input type='button' value='Add' onclick='enroll("+student.id+")'/></td></tr>\n";
			}
			html +="</table>";
			$("#students")[0].innerHTML = html;
		};
		
		$.post(url, callback);
	}else{
		alert("Select a Section!");
	}
}

function enroll(studentId){
	var section = $("#section")[0];
	if(section.selectedIndex != -1 && section.options[section.selectedIndex].value != -1){
		
		var url=base+"enroll="+section.options[section.selectedIndex].value+"&student="+studentId;
		var callback = function(data) {
			var json = parseJSON(data);
			var b = json["results"]["bindings"];
			var success = b[0]["success"]=="true"?"Addition Successful": "Addition Unsuccessful";
			$("#status")[0].innerHTML = success;
			getEnrolledStudents(section.options[section.selectedIndex].value)
		};
		$.post(url, callback);
	}
}

function unEnroll(studentId){
	var section = $("#section")[0];
	if(section.selectedIndex != -1 && section.options[section.selectedIndex].value != -1){
		var url=base+"unenroll="+section.options[section.selectedIndex].value+"&student="+studentId;
		var callback = function(data) {
			var json = parseJSON(data);
			var b = json["results"]["bindings"];
			var success = b[0]["success"]=="true"?"Deletion Successful": "Deletion Unsuccessful";
			$("#status")[0].innerHTML = success;
			getEnrolledStudents(section.options[section.selectedIndex].value)
		};
		$.post(url, callback);
	}
}

function parseJSON(json){
	return eval("("+json+")");
}